resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/KubernetesSubmissionStepsOrchestrator.scala

```diff
@@ -0,0 +1,135 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.kubernetes.submit
+
+import org.apache.spark.SparkConf
+import org.apache.spark.deploy.kubernetes.ConfigurationUtils
+import org.apache.spark.deploy.kubernetes.config._
+import org.apache.spark.deploy.kubernetes.constants._
+import org.apache.spark.deploy.kubernetes.submit.submitsteps.{BaseSubmissionStep, DependencyResolutionStep, DriverKubernetesCredentialsStep, InitContainerBootstrapStep, KubernetesSubmissionStep, PythonStep}
+import org.apache.spark.deploy.kubernetes.submit.submitsteps.initcontainer.InitContainerStepsOrchestrator
+import org.apache.spark.launcher.SparkLauncher
+import org.apache.spark.util.Utils
+
+/**
+ * Constructs the complete list of submission steps to run to deploy the Spark application.
+ */
+private[spark] class KubernetesSubmissionStepsOrchestrator(
+    namespace: String,
+    kubernetesAppId: String,
+    launchTime: Long,
+    mainAppResource: MainAppResource,
+    appName: String,
+    mainClass: String,
+    appArgs: Array[String],
+    additionalPythonFiles: Seq[String],
+    submissionSparkConf: SparkConf) {
+
+  // The resource name prefix is derived from the application name, making it easy to connect the
+  // names of the Kubernetes resources from e.g. Kubectl or the Kubernetes dashboard to the
+  // application the user submitted. However, we can't use the application name in the label, as
+  // label values are considerably restrictive, e.g. must be no longer than 63 characters in
+  // length. So we generate a separate identifier for the app ID itself, and bookkeeping that
+  // requires finding "all pods for this application" should use the kubernetesAppId.
+  private val kubernetesResourceNamePrefix =
+      s"$appName-$launchTime".toLowerCase.replaceAll("\\.", "-")
+  private val jarsDownloadPath = submissionSparkConf.get(INIT_CONTAINER_JARS_DOWNLOAD_LOCATION)
+  private val filesDownloadPath = submissionSparkConf.get(INIT_CONTAINER_FILES_DOWNLOAD_LOCATION)
+  private val dockerImagePullPolicy = submissionSparkConf.get(DOCKER_IMAGE_PULL_POLICY)
+  private val initContainerConfigMapName = s"$kubernetesResourceNamePrefix-init-config"
+
+  def getAllSubmissionSteps(): Seq[KubernetesSubmissionStep] = {
+    val additionalMainAppJar = mainAppResource match {
+      case JavaMainAppResource(resource) if resource != SparkLauncher.NO_RESOURCE =>
+        Option(resource)
+      case _ => Option.empty
+    }
+    val additionalMainAppPythonFile = mainAppResource match {
+      case PythonMainAppResource(resource) if resource != SparkLauncher.NO_RESOURCE =>
+        Option(resource)
+      case _ => Option.empty
+    }
+    val sparkJars = submissionSparkConf.getOption("spark.jars")
+        .map(_.split(","))
+        .getOrElse(Array.empty[String]) ++ additionalMainAppJar.toSeq
```

Comment:
There's no point in having an empty string in the sequence if just not having the element at all would achieve the same effect.